### ISSUE TITLE
Unable to deploy after upgrade to Anguar 17

### ISSUE THREAD

**Issue by mazlano27** [angular/angular#52998]
### Which @angular/* package(s) are the source of the bug?

Don't known / other

### Is this a regression?

Yes

### Description

After upgrading from angular 16 to angular 17 using "ng update", I am not able to deploy my project to the hosting using ng deploy.

### Please provide the exception or error you saw

```true
✔ Server application bundle generation complete.

./server.ts:31:6-15 - Error: export 'default' (imported as 'bootstrap') was not found in './src/main.server' (possible exports: AppServerModule)

./server.ts:56:15-24 - Error: export 'default' (imported as 'bootstrap') was not found in './src/main.server' (possible exports: AppServerModule)

Error: server.ts:9:8 - error TS1192: Module '"/Users/azlan/Desktop/Dev/20-q-gpt/src/main.server"' has no default export.

9 import bootstrap from './src/main.server';
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.0.1
Node: 18.16.0
Package Manager: npm 9.8.1
OS: darwin arm64

Angular: 17.0.3
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router, service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.1
@angular-devkit/build-angular   17.0.1
@angular-devkit/core            17.0.1
@angular-devkit/schematics      17.0.1
@angular/cdk                    17.0.1
@angular/cli                    17.0.1
@angular/fire                   16.0.0
@angular/material               17.0.1
@angular/ssr                    17.0.1
@schematics/angular             17.0.1
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```

### Anything else?

The server.ts file is importing a "bootstrap" module from "./src/main.server". This module does not exist.

**Comment by JoostK** [comment]
A reproduction repro would be necessary to further investigate what is going on here, without it this won't be actionable.

**Comment by HiReception** [comment]
I'm currently having the same issue, I'll see if I can create a reproduction, but given it's an effect of an upgrade (and possibly the result of a repo that has carried through several versions), the problem may not be demonstratable in a newly generated app. For now, I'll provide the details I think are relevant from my experiences so far.

My previous `server.ts` file, on the app running 16.2.0, used `@nguniversal/express-engine`'s `ngExpressEngine` import, and this pattern on the "regular routes" Express handler:

```
import { ngExpressEngine } from '@nguniversal/express-engine';
...
import { AppServerModule } from './src/main.server';

...

server.engine('html', ngExpressEngine({
    bootstrap: AppServerModule,
  }));

...

 // All regular routes use the Universal engine
  server.get('*', (req, res) => {
    res.render(indexHtml, { req, providers: [
      { provide: APP_BASE_HREF, useValue: req.baseUrl },
      { provide: HOST_NAME, useValue: req.get("host")}
    ] });

```

Following the `ng update` migration path, on the rule to replace `@nguniversal` imports with `@angular/ssr` ones, the `server.ts` file was completely replaced with a template file, with the old one copied to a backup:

```> Replace usages of '@nguniversal/' packages with '@angular/ssr'.
RENAME server.ts => server.ts.bak
CREATE server.ts (2358 bytes)
CREATE src/express.tokens.ts (224 bytes)
UPDATE src/app/general/error-page/error-page.component.ts (3310 bytes)
UPDATE package.json (3604 bytes)
✔ Packages installed successfully.
  Migration completed (4 files modified).
```

This new file used a different pattern for "regular routes", relying on a default import from `main.server.ts`, a file which was never changed in the migration and doesn't have a default export (it exports `AppServerModule from './app/app.server.module'`, but this was not an acceptable version for `bootstrap` according to my linter).

```
import { CommonEngine } from '@angular/ssr';
...
import bootstrap from './src/main.server';

...

  const commonEngine = new CommonEngine();

...

// All regular routes use the Angular engine
  server.get('*', (req, res, next) => {
    const { protocol, originalUrl, baseUrl, headers } = req;

    commonEngine
      .render({
        bootstrap,
        documentFilePath: indexHtml,
        url: `${protocol}://${headers.host}${originalUrl}`,
        publicPath: distFolder,
        providers: [
          { provide: APP_BASE_HREF, useValue: baseUrl },
          { provide: RESPONSE, useValue: res },
          { provide: REQUEST, useValue: req }
],
      })
      .then((html) => res.send(html))
      .catch((err) => next(err));
  });
```

Looking through the [example SSR app code](https://github.com/angular/angular/blob/main/aio/content/examples/ssr/src/app/app.config.ts) in the Angular repo, it seems the pattern of setting configuration options for the app (such as injection tokens and providers) have been moved into `app.config.ts` (and `app.server.config.ts`), instead of an `NgModule` declaration in `app.(server.)module.ts`.

It would seem that at least some of the migration logic is assuming that this `config` pattern is being used instead of the `module` one, but I couldn't see any indication on update.angular.io that this was a breaking change. Perhaps some other step in migration was supposed to convert the pattern or handle this use case a different way. For what it's worth, I also received this message at the end of migration relating to the use of Material, but given that it was executed after any `core` or `cdk` migrations, I assume it wouldn't have had an effect on this:

```
> Updates Angular Material to v17.
    Cannot update to Angular Material v17 because the project is using the legacy Material components
    that have been deleted. While Angular Material v16 is compatible with Angular v17, it is recommended
    to switch away from the legacy components as soon as possible because they no longer receive bug fixes,
    accessibility improvements and new features.
    
    Read more about migrating away from legacy components: https://material.angular.io/guide/mdc-migration
    
    Files in the project using legacy Material components:
     - /src/app/live/live.module.ts
    
UPDATE package.json (3604 bytes)
✔ Packages installed successfully.
  Migration completed (1 file modified).
```

I hope this can provide enough information for a lead while I work on a reproduction, and that it wasn't too much of a wall of text 😅

**Comment by HiReception** [comment]
I've created a [reproduction on CodeSandbox](https://codesandbox.io/p/github/HiReception/green-cache-55hm4d/main?file=%2Fnode_modules%2F%40angular%2Fcli%2FLICENSE%3A13%2C32&layout=%257B%2522sidebarPanel%2522%253A%2522GIT%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clp42mcte00073d6lr4yzx1k3%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clp42mctd00033d6lpj7jluuh%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clp42mcte00053d6lkb1wbx67%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clp42mcte00063d6lx688ugyr%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clp42mctd00033d6lpj7jluuh%2522%253A%257B%2522id%2522%253A%2522clp42mctd00033d6lpj7jluuh%2522%252C%2522activeTabId%2522%253A%2522clp432haj00l93d6lcqx5o8g2%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fssr-update-test%252Fpackage.json%2522%252C%2522id%2522%253A%2522clp42ttwv004z3d6ld4cv30p2%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fssr-update-test%252Fsrc%252Fmain.server.ts%2522%252C%2522id%2522%253A%2522clp432haj00l93d6lcqx5o8g2%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522clp42mcte00063d6lx688ugyr%2522%253A%257B%2522id%2522%253A%2522clp42mcte00063d6lx688ugyr%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clp42mcte00053d6lkb1wbx67%2522%253A%257B%2522id%2522%253A%2522clp42mcte00053d6lkb1wbx67%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clp42pira00s43d6lw7ex26au%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clp42pj10001iegijdfbe738g%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clp42pira00s43d6lw7ex26au%2522%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D), using the following steps:

1. Install version 16 of the Angular CLI globally: `npm install -g @angular/cli@16`
2. Create a new app: `ng new ssr-update-test`
3. `cd ssr-update-test` to enter the repo, then add Universal to the app using `ng add @nguniversal/express-engine`
4. Run `ng update`, then `ng update @angular/cli @angular/core` to update to Angular 17
5. Run `npm run dev:ssr` to start a local SSR server, and receive this output instead:
```****************************************************************************************
This is a simple server for use in testing or debugging Angular applications locally.
It hasn't been reviewed for security issues.

DON'T USE IT FOR PRODUCTION!
****************************************************************************************
✔ Browser application bundle generation complete.
✔ Index html generation complete.

Initial Chunk Files | Names         |  Raw Size
vendor.js           | vendor        |   2.29 MB | 
polyfills.js        | polyfills     | 104.06 kB | 
main.js             | main          |  48.04 kB | 
runtime.js          | runtime       |   5.91 kB | 
styles.css          | styles        |   1.22 kB | 

                    | Initial Total |   2.44 MB

Build at: 2023-11-18T13:26:23.252Z - Hash: 978a3943bce6d8db - Time: 34042ms
✔ Server application bundle generation complete.

./server.ts:31:6-15 - Error: export 'default' (imported as 'bootstrap') was not found in './src/main.server' (possible exports: AppServerModule)

./server.ts:56:15-24 - Error: export 'default' (imported as 'bootstrap') was not found in './src/main.server' (possible exports: AppServerModule)

Error: server.ts:9:8 - error TS1192: Module '"/workspaces/workspace/ssr-update-test/src/main.server"' has no default export.

9 import bootstrap from './src/main.server';
         ~~~~~~~~~
```

**Comment by JeanMeche** [comment]
Can you try making the export a default in `main.server.ts` ? 

`export { AppServerModule as default } from './app/app.module.server';
`

**Comment by mazlano27** [comment]
> Can you try making the export a default in `main.server.ts` ?
> 
> `export { AppServerModule as default } from './app/app.module.server'; `

I tried this in my project and it works. I am able to dev:ssr successfully 🤝🏽

**Comment by HiReception** [comment]
> Can you try making the export a default in `main.server.ts` ?
> 
> `export { AppServerModule as default } from './app/app.module.server'; `

I think that's ended up working - I was sure I'd tried that or something very similar, but the server compiles now and seems to work as expected (aside from the additional endpoints I had created in `server.ts` needing to be inserted back in). So I guess the solution would be to add that change to add `as default` into the migration rule.

I'll note as well for the record that the newly created `server.ts` replaced the imports from `fs` and `path` with corresponding ones from `node:fs` and `node:path`, which caused the build error `Cannot find module 'node:fs' or its corresponding type declarations` on each one. That was fixed in my case by removing the `node:` prefixes, so that may also be something to investigate?

In any case, that's resolved my blocker, so thanks for the help!
